src/coreclr/System.Private.CoreLib/src/System/GC.CoreCLR.cs

```diff
@@ -614,6 +615,80 @@ internal static void RegisterMemoryLoadChangeNotification(float lowMemoryPercent
             }
         }
 
+        [Flags]
+        internal enum FinalizerActionKind
+        {
+            None = 0,
+            ExecuteNoGCCallback = 1,
+            DetachNoGCCallback = 2,
+        }
+
+        internal enum EnableNoGCRegionCallbackStatus
+        {
+            Succeed,
+            NotStarted,
+            InsufficientBudget,
+        }
+
+        private static Action? s_NoGCCallback;
+
+        [UnmanagedCallersOnly]
+        private static unsafe void FinalizerAction(FinalizerActionKind finalizerActionKind)
+        {
+            Action? callback = s_NoGCCallback;
+            if ((finalizerActionKind & FinalizerActionKind.DetachNoGCCallback) != 0)
+            {
+                s_NoGCCallback = null;
+            }
+            if ((finalizerActionKind & FinalizerActionKind.ExecuteNoGCCallback) != 0)
+            {
+                callback?.Invoke();
+            }
+        }
+
+        /// <summary>
+        /// Register a callback to be invoked when we allocated a certain amount of memory in the no GC region.
+        /// <param name="totalSize">The total size of the no GC region. Must be a number > 0 or an ArgumentOutOfRangeException will be thrown.</param>
+        /// <param name="callback">The callback to be executed when we allocated a certain amount of memory in the no GC region..</param>
+        /// <exception cref="System.ArgumentOutOfRangeException"> The <paramref name="totalSize"/> argument is less than or equal to 0.</exception>
+        /// <exception cref="System.ArgumentNullException">The <paramref name="callback"/> argument is null.</exception>
+        /// <exception cref="InvalidOperationException"><para>The GC is not currently under a NoGC region.</para>
+        /// <para>-or-</para>
+        /// <para>Another callback is already registered.</para>
+        /// <para>-or-</para>
+        /// <para>The <paramref name="totalSize"/> exceeds the size of the No GC region.</para>
+        /// <para>-or-</para>
+        /// <para>We failed to withheld memory for the callback before of already made allocation.</para>
+        /// </exception>
+        /// </summary>
+        public static unsafe void RegisterNoGCRegionCallback(long totalSize, Action callback)
+        {
+            ArgumentOutOfRangeException.ThrowIfNegativeOrZero(totalSize);
+            ArgumentNullException.ThrowIfNull(callback);
+
+            if (Interlocked.CompareExchange(ref s_NoGCCallback, callback, null) != null)
+            {
+                throw new InvalidOperationException(SR.InvalidOperationException_NoGCRegionCallbackAlreadyRegistered);
+            }
+
+            _RegisterFinalizerAction(&FinalizerAction);
+            EnableNoGCRegionCallbackStatus status = _EnableNoGCRegionCallback(totalSize);
+            if (status != EnableNoGCRegionCallbackStatus.Succeed)
```

Comment:
I am wondering about implications of throwing exception here on how to use this API reliably. It sounds like that users will want to wrap calls of this API in try/catch blocks to protect against cases like machine being momentarily overloaded and the process making too much progress on other threads and exceeding the budget already or even finishing the no GC region.

Should this be highlighted in the docs?

Reply:
(Or reconsider how the API should behave in these situations to avoid this failure pattern.)

Reply:
The API documentation does include the list of conditions where this API call might fail.